### ISSUE TITLE
Getting old profile when retrieving different users

### ISSUE THREAD

**Issue by rfossella** [node-saml/node-saml#379]
Hi,

Using  
"passport": "^0.5.3",
"passport-saml": "^3.2.4",
"express": "^4.18.2",
"express-session": "^1.17.3",
node 16.20

Browsers 
Chrome Version 129.0.6668.90 (Official Build) (64-bit)
Edge Version 129.0.2792.89 (Official build) (64-bit)

**I THOUGHT THIS ISSUE WAS ONLY IN CHROME, BUT IT OCCURS IN EDGE TOO**

---
I have a passport-saml set up whereby I have to support **multiple** sets of credentials - with IdP MS Entra ID

Because of this, in my /login route I dynamically get the creds and create new SamlStrategy

```
router.post('/login', (req, res, next) => 
{
    const username = req.body.username;
    const role= req.body.role; 
    const action = req.body.action;

passport.use(new SamlStrategy(getSAMLCredentials(username), 
		(profile, done) =>
		{
			console.log('SAML Profile:', profile);
		}))

// Initiate SAML login
passport.authenticate('saml')(req, res, next);

```
I have three (3) users.  When I **first** login USER1 I properly get prompted through to MS Entra to select user name, then password.
I then correctly hit callback code with a valid profile.   I then correctly hit my login/callback route via MS Entra.  I do my req.login(user..) and user is successfully logged into my application.

I then logout that user using:
```
req.logOut();
req.session.destroy(err => 
{
if (err) {
		console.error('Failed to destroy session:', err);
} else {
		console.log('Session successfully destroyed');
}
	});
	
res.clearCookie('connect.sid', { path: '/' });
```
User is logged out and session shows to be destroyed.

I go back to my login screen.

When I then enter USER2 it appears that I don't get back out to MS Entra to select THAT user name and instead I am still holding onto the OLD/LAST profile from USER1.  I have extra logic to make sure that _entered_ user name matches the _profile_ user name  - this certainly fails because again I am still holding USER1 profile.

If I manually logout of MS Entra - either from browser or programmatically, and then enter USER2, I get a new (correct) prompt for Entra and new fresh profile.  But the cycle continues w/USER3, etc.

As a work-around, I added the following:

```
passport.use(new SamlStrategy({
				...getSAMLCredentials(username),
				forceAuthn: true
			  }, (profile, done) =>
					{
						console.log('SAML Profile:', profile); 
                                     })
```

This is now FORCING a SAML (SSO) request every single time (correctly, based on documentation) but this defeats the purpose of a Single-Sign-On implementation where you only want to have the user logged in once - unless you explicitly log out of IdP.

**Expected behavior**
So, my expectation is:

1. If I FIRST-TIME login USER1, I get the whole Entra user/password check and user gets into application.
2. I then log USER1 out (of application, not Entra).  
3. If I login USER1 again he should go directly into application as he SSO'd already.
4. If I then FIRST-TIME login USER2, I should be again directed to Entra user/pwd, get a new profile back - USER2 gets logged in
5. I log USER2 out of application only
6. Back to USER1 - should go right in... etc.

So again, I am stuck on the thing where I seem to holding onto an old profile even though I am destroying the session , req.logout(), etc.

Is there some else I should be doing w/respect to passport-saml to get clean profiles?  Is this something strange going on w/IdP?  Browsers and cache?

**Comment by rfossella** [comment]
Thanks for your valuable feedback. I agree - this should have been a discussion. Maybe I should close this and create discussion thread?

Interesting about the deprecation.  I actually went to the Passport home page (https://www.passportjs.org/packages/passport-saml/) and followed their current instructions and ended up w/v3.2.4.  I will certainly install @node-saml/passport-saml

Regarding the SLO - you are correct.  When I programmatically trigger the IdP SLO (along with application logout) it will work fine.  However, when the user logs back in they have to re-login to IdP, which is something we wanted to avoid. 

I will read through the SLO link you gave and also the use of MultiSamlStrategy which may help.

Appreciate  your feedback!

**Comment by rfossella** [comment]
An issue arose when using the new version whereby it complains about **idpCert not found** - even though using same syntax - i.e.

```
let samlCredentials = getSAMLCredentials(username);
passport.use(new SamlStrategy({...samlCredentials, forceAuthn: false}, 
(profile, done) =>
{
console.log('SAML Profile:', profile); 
})
```

Works fine with **passport-saml**;  fails with **@node-saml/passport-saml**, even though the doc on new page shows same syntax.  My certs are exactly the same.

I read something on stack about needing, wantAssertionsSigned:  and wantAuthnResponseSigned: but had no effect.

I 
```
return {
		entryPoint,
		issuer,
		cert,
		callbackUrl,
		logoutUrl,
		wantAssertionsSigned: false,
		wantAuthnResponseSigned: false
		}
```

Trying the find a syntactical difference in docs as to why this would fail w/new package.

**Comment by rfossella** [comment]
Firstly, thank you for the reference.  I had read that and focused on the first example, seeing the same syntax using "cert".  I did not make the connection referencing ADFS and using "idpCert."  This has since worked for me.


> You tried to resolve "idpCert not found" with those config parameters? I.e. copy pasted this and that to security related component configuration without taking time to understand implications

The stack comments I read specifically pointed to certificate issues.  Reading through lots of posts.  So yes, I must confess, while it certainly seemed unrelated, sadly I was _grasping_ here to see if that config option needed. 

Thanks again for your help.

**Comment by rfossella** [comment]
I was looking for #378 which appears related to a new error I am seeing "Error: Invalid document signature\n    at SAML.validatePostResponseAsync "

Using my previous use case, this occurs when I initially choose USER1, then follow up by using USER2. The second user fails with that error.

**Comment by markstos** [comment]
I opened an issue for the Passport website to have them reference the right version. 

https://github.com/passport/www.passportjs.org/issues/112

**Comment by rfossella** [comment]
srd90 - More background on the use of a username.   When a user enters our system we do not initially know what Idp credentials (company) they are associated with.  We are a multi-tenant system.  So user user1@xyz.com enters their name and presses enter;  we lookup user1@xyz.com and get the idp creds that are associated with xyz.com.  That's how the unique idp creds are gotten.

Each of those users are potentially associated with separate MS Entra accounts.  So when MS pops up their initial screen you choose user1@xyz.com.  User2 maybe user2@abc.com with completely different idp creds.   Again, mutli-tenant system w/different idp creds.

Note: this was tested 100's of times w/**passport-saml** package, with no cert errors not the latest "document signature" errors.  The only issue is what my original post stated, that  sometimes I would be holding onto an old profile.

That said, these new issues came up with **@node-saml/node-saml** installed.  You pointed me to the use of **idpcert** instead of **cert** for my MSEntra IDP.  FYI: Our users/companies may not (probably won't) **all** use MSEntra.  That certainly fixed the cert error when I login the first user.  

All my other logic is the same.  I trusted that the new version would be a little more seamless.

I hope I cleared up the use of username and our multi-tenant system.

I also saw their is a multistrategy option available.  I planned to look into that once i had the new version working like the old version - obviously this where my snag is.

I will take a look at some of your other suggestions in you post.

**Comment by rfossella** [comment]
I read through all the valuable comments.  The clients that use our application must be white-listed so even the first iteration, not requiring assertion level signatures, was satisfactory.

I did loosen up the restrictions using the options, and I'm no longer getting the signature error.  _Part_ of the issue with this is instructing our clients to make the correct settings in their respective Idp - not that it is necessarily a bad thing.
```
wantAssertionsSigned: false,
wantAuthnResponseSigned: false
```
So I at least have a stable working QA version using latest release.

FYI: the "old" profile issue remains, but as you stated, it certainly could be related to other things outside of passport-saml.  Technically speaking, this issue primarily comes up in my testing suite since I have about six different users that I test with that are in and out of "signed-in status."  Most, if not all of our users, use a single client login user id.

That said, I did add an _optional_ UI/UX option to the app that will let the user know there are multiple "signed-in users" trying to access same app w/same IdP credentials.  In this case I give them the option to SLO the _problematic_ user.  I discussed w/Product Management and also forcing a SLO when the user _only_ meant to sign out of local app is not what they want.  They are aware of the possible snag.  If I figure out anything more about that issue I will share in the forum.

Thank you again for all your help with this.  Very much appreciated.
